ssl/ech/ech_helper.c

```diff
@@ -11,5 +11,136 @@
 #include <openssl/ech.h>
 #include "../ssl_local.h"
 #include "ech_local.h"
+#include "internal/ech_helpers.h"
 
-/* TODO(ECH): move code that's used by internals and test here */
+/* TODO(ECH): move more code that's used by internals and test here */
+
+/* used in ECH crypto derivations (odd format for EBCDIC goodness) */
+/* "tls ech" */
+static const char OSSL_ECH_CONTEXT_STRING[] = "\x74\x6c\x73\x20\x65\x63\x68";
+
+/*
+ * Given a SH (or HRR) find the offsets of the ECH (if any)
+ * sh is the SH buffer
+ * sh_len is the length of the SH
+ * exts points to offset of extensions
+ * echoffset points to offset of ECH
+ * echtype points to the ext type of the ECH
+ * return 1 for success, other otherwise
+ *
+ * Offsets are returned to the type or length field in question.
+ * Offsets are set to zero if relevant thing not found.
+ *
+ * Note: input here is untrusted!
+ */
+int ossl_ech_get_sh_offsets(const unsigned char *sh, size_t sh_len,
+                            size_t *exts, size_t *echoffset,
+                            uint16_t *echtype)
+{
+    unsigned int elen = 0, etype = 0, pi_tmp = 0;
+    const unsigned char *pp_tmp = NULL, *shstart = NULL, *estart = NULL;
+    PACKET pkt;
+    size_t extlens = 0;
+    int done = 0;
+#ifdef OSSL_ECH_SUPERVERBOSE
+    size_t echlen = 0; /* length of ECH, including type & ECH-internal length */
+    size_t sessid_offset = 0, sessid_len = 0;
+#endif
+
+    if (sh == NULL || sh_len == 0 || exts == NULL || echoffset == NULL
+        || echtype == NULL)
+        return 0;
+    *exts = *echoffset = *echtype = 0;
+    if (!PACKET_buf_init(&pkt, sh, sh_len))
+        return 0;
+    shstart = PACKET_data(&pkt);
+    if (!PACKET_get_net_2(&pkt, &pi_tmp))
+        return 0;
+    /*
+     * TODO(ECH): we've had a TLSv1.2 test in the past where we add an
+     * ECH to a TLSv1.2 CH to ensure server code ignores that properly.
+     * We might or might not keep that, if we don't then the test below
+     * should allow TLSv1.3 only.
+     */
+    /* if we're not TLSv1.2+ then we can bail, but it's not an error */
+    if (pi_tmp != TLS1_2_VERSION && pi_tmp != TLS1_3_VERSION)
+        return 1;
```

Comment:
I think yes, for now. There's been some recent IETF TLS WG discussion that a new DTLSv1.3 RFC is needed to resolve some ambiguities in the protocol (found by boringssl folk). So far, there no change related to ECH, but that could change, perhaps most likely a documentation change rather than one affecting the code. But in any case, it may be better to just do ECH for TLS now, and then return to consider QUIC and DTLS. (Also, I've done no testing whatsoever for DTLS nor looked at code really.)